### DISCUSSION TITLE
Selecting subset of rows of many to many relation with conditional LEFT JOIN while using QueryBuilder

### DISCUSSION THREAD

**Discussion by pawelsmolenski** [mikro-orm/mikro-orm#4643] [Q&A]
I wanted to clarify JOIN conditions behavior and/or its proper usage before opening an issue:

I have two entities with many to many relationship between them (without explicitly defined pivot table):

```typescript
export class TableEntity {
    @PrimaryKey({ defaultRaw: "uuid_generate_v4()", type: "uuid" })
    id!: string;

    @ManyToMany({
        entity: () => ReservationEntity,
        inversedBy: (reservation: ReservationEntity) => reservation.tables,
    })
    reservations: Collection<ReservationEntity> = new Collection<ReservationEntity>(this);
}
```

```typescript
export class ReservationEntity {
    @PrimaryKey({ defaultRaw: "uuid_generate_v4()", type: "uuid" })
    id!: string;

    @Property({ type: "timestamp with time zone" })
    end: Date;

    @Property({ type: "timestamp with time zone" })
    start: Date;

    @ManyToMany({
        entity: () => TableEntity,
        mappedBy: (table: TableEntity) => table.reservations,
    })
    tables: Collection<TableEntity> = new Collection<TableEntity>(this);
}
```

Entities represent restaurant tables and their reservations in time. I want to query for all tables with left joined subset of reservations which take place within the given time period:

```typescript
    const tables: Array<TableEntity> = await em.createQueryBuilder(TableEntity, "t")    
        .select("t.*")
        .leftJoinAndSelect("t.reservations", "r", {
            $and: [
                { "r.start": { $lte: input.end } },
                { "r.end": { $gte: input.start } },
            ]
        })
        .getResult();
```

It seems that this query is not working as expected, because the join condition is applied to the pivot table (which is unknown to me) instead of the target entity of the relation:

```sql
select "t".*, "r"."id" as "r__id", "r"."end" as "r__end", "r"."start" as "r__start" 
    from "tables" as "t"
    left join "tables_reservations" as "t1" 
        on "t"."id" = "t1"."table_entity_id" 
            and ("r"."start" <= '2023-08-21T11:22:53.982Z' and "r"."end" >= '2023-08-21T11:22:53.982Z') 
    left join "reservations" as "r" 
        on "t1"."reservation_entity_id" = "r"."id"
```

Should this query be constructed differently or can we consider this behavior a bug?

Dependency | Version
-- | --
typescript | 5.1.6
mikro-orm | 5.7.14
postgresql (pg) | 14

**Comment by B4nan** [comment]
> It seems that this query is not working as expected, because the join condition is applied to the pivot table (which is unknown to me) instead of the target entity of the relation:

Why do you say so? The pivot table is aliased to `t1`, not to `r`, that query seems to be correct, the condition is applied to the target entity and not the pivot table, or what am I missing?

edit: oh now I see it, yeah that is a bug for sure, moved to the issues (#4644)
